src/mpi-base.jl

```diff
@@ -156,6 +120,54 @@ function Comm_size(comm::Comm)
     Int(size[])
 end
 
+function Type_create_struct{T <: Any}(::Type{T})  # <: Any effectively 
+                                                  # limits T to being a Type
+
+  @assert isbits(T)
+  # get the data from the type
+  fieldtypes = T.types
+  offsets = fieldoffsets(T)
+  nfields = Cint(length(fieldtypes))
+
+  # put data in MPI format
+  blocklengths = ones(Cint, nfields)
+  displacements = zeros(Cptrdiff_t, nfields)  # size_t == MPI_Aint ?
+  types = zeros(Cint, nfields)
+  for i=1:nfields
+    displacements[i] = offsets[i]
+    types[i] = mpitype(fieldtypes[i])
+  end
+
+  # create the datatype
+  newtype_ref = Ref{Cint}()
+  flag = Ref{Cint}()
+  ccall(MPI_TYPE_CREATE_STRUCT, Void, (Ptr{Cint}, Ptr{Cint}, Ptr{Cptrdiff_t}, 
+        Ptr{Cint}, Ptr{Cint}, Ptr{Cint}), &nfields, blocklengths, displacements, types, 
+        newtype_ref, flag)
+
+  if flag[] != 0
+    println(STDERR, "Warning: MPI_TYPE_CREATE_STRUCT returned non-zero exit stats")
+  end
+
+  # commit the datatatype
+  flag2 = Ref{Cint}()
+
+  ccall(MPI_TYPE_COMMIT, Void, (Ptr{Cint}, Ptr{Cint}), newtype_ref, flag2)
+
+  if flag2[] != 0
+    println(STDERR, "Warning: MPI_TYPE_COMMIT returned non-zero exit status")
+  end
+
+  # add it to the dictonary of known types
+  mpitype_dict[T] = newtype_ref[]
```

Comment:
If this function is called multiple times for the same Julia type, then it should execute only once. The other calls should do nothing.